dig.go

```diff
@@ -275,3 +307,85 @@ func detectCycles(n node, graph map[reflect.Type]node, path []reflect.Type, seen
 	}
 	return nil
 }
+
+// Param is embedded inside structs to opt those structs in as Dig parameter
+// objects.
+//
+// TODO usage docs
+type Param struct{}
+
+var _ parameterObject = Param{}
+
+// Param is the only instance of parameterObject.
+func (Param) parameterObject() {}
+
+// Users embed the Param struct to opt a struct in as a parameter object.
+// Param implements this interface so the struct into which Param is embedded
+// also implements this interface. This provides us an easy way to check if
+// something embeds Param without iterating through all its fields.
+type parameterObject interface {
+	parameterObject()
+}
+
+// Returns dependencies introduced by a parameter object.
+func getParameterDependencies(t reflect.Type) ([]reflect.Type, error) {
+	for t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	var deps []reflect.Type
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if f.PkgPath != "" {
+			continue // skip private fields
+		}
+
+		// Skip the embedded Param type.
+		if f.Anonymous && f.Type == _paramType {
+			continue
+		}
+
+		// THe user added a parameter object as a dependency. We don't recurse
+		// /yet/ so let's try to give an informative error message.
+		if f.Type.Implements(_parameterObjectType) {
```

Comment:
I've tried private implementations and it looks very confusing:
foo/foo.go:
```golang
package foo

type A struct {}

func (A) private() {
	println("foo")
}
```
main.go
```golang
package main

import (
	"fmt"
	"tmp/foo"
	"reflect"
)

type B struct{foo.A}
type private interface {
	private()
}

func main() {
	b := reflect.ValueOf(B{})
	v := (*private)(nil)
	fmt.Println(reflect.TypeOf(v).Elem())
	fmt.Println(b.Type().Implements(reflect.TypeOf(v).Elem()))
}
```
the output is:
```
main.private
true
```
May be I am doing something wrong, but it seems there could be conflicts if a type has parameterObject() method..